src/System.Security.Cryptography.Primitives/src/System/Security/Cryptography/CryptoStream.cs

```diff
@@ -147,9 +151,36 @@ public override void SetLength(long value)
             throw new NotSupportedException(SR.NotSupported_UnseekableStream);
         }
 
+        public override async Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            CheckReadArguments(buffer, offset, count);
+
+            // To avoid a race with a stream's position pointer & generating race 
+            // conditions with internal buffer indexes in our own streams that 
+            // don't natively support async IO operations when there are multiple 
+            // async requests outstanding, we will block the application's main
+            // thread if it does a second IO request until the first one completes.
+
+            SemaphoreSlim semaphore = AsyncActiveSemaphore;
+            await semaphore.WaitAsync().ForceAsync();
+            try
+            {
+                return await ReadAsyncCore(buffer, offset, count, cancellationToken, useAsync: true);
```

Comment:
Ideally it would be, if for no other reason than it'll avoid questions about why it's not.  However, it's likely not necessary, as the ForceAsync Ati's using on the previous line will guarantee that this code is running asynchronously from the caller on a thread pool thread, where there should be a default context/scheduler.